File: api/src/main/java/fr/huiitre/tools/modules/core/notification/application/usecase/MarkNotificationsAsReadUseCase.java
package fr.huiitre.tools.modules.core.notification.application.usecase;

import java.util.List;

import org.springframework.stereotype.Service;

import fr.huiitre.tools.modules.core.role.domain.RoleCode;
import fr.huiitre.tools.modules.core.security.application.ports.CurrentUserProvider;
import fr.huiitre.tools.modules.core.security.application.usecase.SecuredUseCase;
import fr.huiitre.tools.modules.core.notification.application.port.NotificationRepository;

@Service
public class MarkNotificationsAsReadUseCase implements SecuredUseCase {

    private final NotificationRepository notificationRepository;
    private final CurrentUserProvider currentUserProvider;

    public MarkNotificationsAsReadUseCase(NotificationRepository notificationRepository, CurrentUserProvider currentUserProvider) {
        this.notificationRepository = notificationRepository;
        this.currentUserProvider = currentUserProvider;
    }

    @Override
    public RoleCode requiredRole() {
        return RoleCode.READ_ONLY;
    }

    public void execute(List<Long> notificationIds) {
        Long userId = Long.parseLong(currentUserProvider.getCurrentUserId());
        notificationRepository.markAsRead(userId, notificationIds);
    }
}

File: api/Modules/Core/Notifications/Application/Usecases/MarkNotificationsAsReadUseCase.cs
using Tools.Api.Modules.Core.Notifications.Application.Ports;
using Tools.Api.Modules.Core.Security.Application.Services;
using Tools.Api.Modules.Core.Security.Application.Usecases;

namespace Tools.Api.Modules.Core.Notifications.Application.Usecases;

// Marque comme lues les notifications de l'appelant.
public sealed class MarkNotificationsAsReadUseCase(
    UseCaseAuthorizer authorizer,
    INotificationRepository notificationRepository) : SecuredUseCase(authorizer)
{
    // `notificationIds` nul : tout est marqué lu. L'appelant ne peut agir que sur ses propres
    // lignes — le filtre porte sur `CurrentUser.UserId`, jamais sur une valeur reçue du client.
    public Task Execute(IReadOnlyCollection<long>? notificationIds)
    {
        return notificationRepository.MarkAsReadAsync(CurrentUser.UserId, notificationIds);
    }
}
